smart-contract-feature-development.md

```diff
@@ -0,0 +1,94 @@
+
+Feature branches result in a lot of effort spent keeping the feature branch up to date. We want to have modular code
+that enables us to easily compose features together such that we can get code mainlined early and toggle it off when
+running in production.
+
+# Problem Statement + Context
+
+We want to move away from a world with long lived feature branches.
+To do so, we need to have some flexibility when it comes to composition of features.
+This is easy in off-chain code as the cost is minimal for toggling off certain features but still having them in the source code.
+This is not the case for Solidity. This document exists to outline how to think about composition of features for Solidity code.
+When the feature is ready to be shipped, the feature specific contract should be merged into the base contract.
+
+# Alternatives Considered
+
+## Copy and Paste Implementation
+
+This was the solution for the fault proofs project, which worked well until the custom gas token feature was introduced. Now the contracts
+are in a state such that the custom gas token feature needs to be ported to the `OptimismPortal2` before the `OptimismPortal` and the
+`OptimismPortal2` can be merged. This may be a good solution for security critical code as we do not rely on the correctness
+of solidity inheritance and all of the logic lives in a single contract, making it easier to audit.
+
+# Proposed Solution
+
+Below includes general guidelines for how to develop solidity feature contracts that are meant to be merged into mainline at
+some point in the future.
+
+### Composition through Inheritance
+
+Solidity allows contacts to [inherit](https://docs.soliditylang.org/en/v0.8.17/contracts.html#inheritance) other contracts.
+The functionality of the contract can then be extended or overridden.
+For in development features that modify core contracts already running in production,
+a new contract should be created that inherits the core contract and all new functionality should be added to the new contract.
+The deploy script should be able to parse some sort of config and optionally deploy the feature enabled version of the contract.
+
+### Handling Inheritance
+
+Inherit the contract that needs new functionality. It is very easy to add net new functions. If certain functions need to be modified,
+then the original contract may need to have the `virtual` keyword added in places. The `virtual` keyword should only be added in
+places where it must be added and nowhere in addition.
+
+### Handling Semver
+
+The parent contract should update its `version` getter to be a real function. The child contract should implement a `version` getter like so:
+
+```solidity
+function version() public override returns (string) {
+  return string.concat(super.version(), "+feature-name");
+}
+```
+
+The `feature-name` should be some identifier that uniquely identifies the feature. It could also be a hardfork name if the release is tied to a hard fork.
+
+### Handling Storage
```

Comment:
One final idea is giving each feature contract a dedicated storage contract, instead of using the existing contract. I don't love this idea because it's a bit clunky, but it would work and keep all storage isolated, so wanted to mention it for completeness. Example:

```solidity
contract OptimismPortal {
  // --- snip ---
}

contract OptimismPortalFeatA is OptimismPortal {
  OptimismPortalFeatAStorage featAStorage; 

  function initialize() external initializer {
    // Need to avoid redeployment on future initialization calls, one way is use
    // create2/create3 with a constant salt, and only deploy if there is no code at
    // the expected address.
    featAStorage = new OptimismPortalFeatAStorage();
  }

  function setX(uint256 x) external {
    featAStorage.setX(x);
  }

  function getX() external returns (uint256) {
    return featAStorage.x();
  }
}

contract OptimismPortalFeatAStorage {
  address public immutable parent;
  uint256 public x;

  constructor() {
    parent = msg.sender;
  }

  function setX(uint256 _x) external {
    require(msg.sender == parent, "Unauthorized");
    x = _x;  
  }
}
```

Reply:
I actually like this solution a lot for the predeploys, at least for chain specific storage. In practice the `initialize` wouldn't need to deploy a contract, it could just use a known predeploy.

I don't love this for the L1 contracts, but it does create an abstraction that works for storage